libsolidity/analysis/OverflowChecker.cpp

```diff
@@ -0,0 +1,220 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author Rhett <roadriverrail@gmail.com>
+ * @date 2017
+ * Overflow analyzer and checker.
+ */
+
+#include <libsolidity/analysis/OverflowChecker.h>
+#include <libsolidity/analysis/OverflowBounds.h>
+#include <libsolidity/analysis/IntegerOverflowBounds.h>
+#include <memory>
+#include <boost/range/adaptor/reversed.hpp>
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/inlineasm/AsmAnalysis.h>
+#include <libsolidity/inlineasm/AsmAnalysisInfo.h>
+#include <libsolidity/inlineasm/AsmData.h>
+#include <libsolidity/interface/ErrorReporter.h>
+#include <iostream>
+#include <typeinfo>
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+using namespace boost::multiprecision;
+
+bool OverflowChecker::checkOverflow(ASTNode const& _contract)
+{
+	try
+	{
+		_contract.accept(*this);
+	}
+	catch (FatalError const&)
+	{
+		// We got a fatal error which required to stop further checking, but we can
+		// continue normally from here.
+		if (m_errorReporter.errors().empty())
+			throw; // Something is weird here, rather throw again.
+	}
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
+}
+
+bool OverflowChecker::visit(Assignment const& _assignment)
+{
+	_assignment.rightHandSide().accept(*this);
+	auto const identifier = dynamic_cast<Identifier const*>(&_assignment.leftHandSide());
+	if (identifier)
+	{
+		shared_ptr<OverflowBounds>bounds = m_overflowInfo[&_assignment.rightHandSide()];
+		auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
```

Comment:
This seems like a good place to ask the question.  I feel like I'm making a lot of `dynamic_casts` and I feel like it might be a sign of awkward code.  Here, it's obviously needed, but elsewhere, I'm often doing it just to force type compliance, and something feels wrong about that.